astropy/cosmology/flrw.py

```diff
@@ -114,52 +115,33 @@ class FLRW(Cosmology):
-                   unit="km/(s Mpc)")
-    Om0 = Parameter(doc="Omega matter; matter density/critical density at z=0.")
-    Ode0 = Parameter(doc="Omega dark energy; dark energy density/critical density at z=0.")
+                   unit="km/(s Mpc)", fvalidate="scalar")
+    Om0 = Parameter(doc="Omega matter; matter density/critical density at z=0.",
+                    fvalidate="non-negative")
+    Ode0 = Parameter(doc="Omega dark energy; dark energy density/critical density at z=0.",
+                     fvalidate="float")
     Tcmb0 = Parameter(doc="Temperature of the CMB as `~astropy.units.Quantity` at z=0.",
-                      unit="Kelvin", fmt="0.4g")
-    Neff = Parameter(doc="Number of effective neutrino species.")
+                      unit="Kelvin", fmt="0.4g", fvalidate="scalar")
+    Neff = Parameter(doc="Number of effective neutrino species.", fvalidate="non-negative")
     m_nu = Parameter(doc="Mass of neutrino species.",
                      unit="eV", equivalencies=u.mass_energy(), fmt="")
     Ob0 = Parameter(doc="Omega baryon; baryonic matter density/critical density at z=0.")
 
     def __init__(self, H0, Om0, Ode0, Tcmb0=0.0*u.K, Neff=3.04, m_nu=0.0*u.eV,
                  Ob0=None, *, name=None, meta=None):
         super().__init__(name=name, meta=meta)
-        cls = self.__class__
-
-        # all densities are in units of the critical density
-        self._Om0 = float(Om0)
-        if self._Om0 < 0.0:
-            raise ValueError("Matter density can not be negative")
-        self._Ode0 = float(Ode0)
-        if Ob0 is not None:
-            self._Ob0 = float(Ob0)
-            if self._Ob0 < 0.0:
-                raise ValueError("Baryonic density can not be negative")
-            if self._Ob0 > self._Om0:
-                raise ValueError("Baryonic density can not be larger than "
-                                 "total matter density")
-            self._Odm0 = self._Om0 - self._Ob0
-        else:
-            self._Ob0 = None
-            self._Odm0 = None
-
-        self._Neff = float(Neff)
-        if self._Neff < 0.0:
-            raise ValueError("Effective number of neutrinos can "
-                             "not be negative")
 
-        # Tcmb may have units
-        self._Tcmb0 = Tcmb0 << cls.Tcmb0.unit
-        if not self._Tcmb0.isscalar:
-            raise ValueError("Tcmb0 is a non-scalar quantity")
+        # Assign (and validate) Parameters
+        self.H0 = H0
+        self.Om0 = Om0
+        self.Ode0 = Ode0
+        self.Tcmb0 = Tcmb0
+        self.Neff = Neff
+        # self.m_nu = m_nu  # set later
+        self.Ob0 = Ob0  # (must be after Om0)
 
-        # Hubble parameter at z=0, km/s/Mpc
-        self._H0 = H0 << cls.H0.unit
-        if not self._H0.isscalar:
-            raise ValueError("H0 is a non-scalar quantity")
+        # Derived quantities
+        self._Odm0 = None if Ob0 is None else (self._Om0 - self._Ob0)
```

Comment:
This way of doing derived parameters seems nicer than the detailed calculation done below.